File: chrome/android/java/src/org/chromium/chrome/browser/download/service/DownloadTaskScheduler.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.download.service;

import org.chromium.base.ContextUtils;
import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.components.background_task_scheduler.BackgroundTaskScheduler;
import org.chromium.components.background_task_scheduler.BackgroundTaskSchedulerFactory;
import org.chromium.components.background_task_scheduler.TaskIds;
import org.chromium.components.background_task_scheduler.TaskInfo;

import java.util.concurrent.TimeUnit;

/**
 *
 */
@JNINamespace("download::android")
public class DownloadTaskScheduler {
    @CalledByNative
    private static void scheduleDownloadTask(
            boolean requiresCharging, boolean requiresUnmeteredNetwork) {
        BackgroundTaskScheduler scheduler = BackgroundTaskSchedulerFactory.getScheduler();
        TaskInfo taskInfo = TaskInfo.createOneOffTask(TaskIds.DOWNLOAD_SERVICE_JOB_ID,
                                            DownloadBackgroundTask.class, 0)
                                    .setRequiredNetworkType(requiresUnmeteredNetwork
                                                    ? TaskInfo.NETWORK_TYPE_UNMETERED
                                                    : TaskInfo.NETWORK_TYPE_ANY)
                                    .setRequiresCharging(requiresCharging)
                                    .setUpdateCurrent(true)
                                    .setIsPersisted(true)
                                    .build();
        scheduler.schedule(ContextUtils.getApplicationContext(), taskInfo);
    }

    @CalledByNative
    private static void scheduleCleanupTask(long keepAliveTimeMs) {
        BackgroundTaskScheduler scheduler = BackgroundTaskSchedulerFactory.getScheduler();
        TaskInfo taskInfo = TaskInfo.createOneOffTask(TaskIds.DOWNLOAD_CLEANUP_JOB_ID,
                                            DownloadBackgroundTask.class, keepAliveTimeMs,
                                            keepAliveTimeMs + TimeUnit.HOURS.toMillis(1))
                                    .setRequiresCharging(true)
                                    .setIsPersisted(true)
                                    .build();
        scheduler.schedule(ContextUtils.getApplicationContext(), taskInfo);
    }

    /*@CalledByNative
    private static void cancelDownloadTask() {
        BackgroundTaskScheduler scheduler = BackgroundTaskSchedulerFactory.getScheduler();
        scheduler.cancel(ContextUtils.getApplicationContext(), TaskIds.DOWNLOAD_SERVICE_JOB_ID);
    }*/
}

File: chrome/browser/android/download/service/download_task_scheduler.cc
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/android/download/service/download_task_scheduler.h"

#include "base/android/jni_android.h"
#include "jni/DownloadTaskScheduler_jni.h"

namespace download {
namespace android {

DownloadTaskScheduler::DownloadTaskScheduler() = default;

DownloadTaskScheduler::~DownloadTaskScheduler() = default;

void DownloadTaskScheduler::ScheduleDownloadTask(
    bool require_charging,
    bool require_unmetered_network) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_DownloadTaskScheduler_scheduleDownloadTask(env, require_charging,
                                                  require_unmetered_network);
}

void DownloadTaskScheduler::ScheduleCleanupTask(long keep_alive_time_ms) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_DownloadTaskScheduler_scheduleCleanupTask(env, keep_alive_time_ms);
}

}  // namespace android
}  // namespace download
